client.go

```diff
@@ -320,11 +325,171 @@ func (client *client) GetOffset(topic string, partitionID int32, time int64) (in
+	return client.tryCommitOffset(consumerGroup, topic, partitionID, offset, metadata, 5)
+}
+
+func (client *client) tryCommitOffset(consumerGroup string, topic string, partitionID int32, offset int64, metadata string, attemptsRemaining int) error {
+	var err error
+
+	broker, err := client.getCoordinator(consumerGroup, 10)
+	if err != nil {
+		return err
+	}
+
+	offsetCommitRequest := new(OffsetCommitRequest)
+	offsetCommitRequest.ConsumerGroup = consumerGroup
+	offsetCommitRequest.AddBlock(topic, partitionID, offset, ReceiveTime, metadata)
+
+	response, err := broker.CommitOffset(offsetCommitRequest)
+	if err != nil {
+		goto RetryCommitHandler
+	}
+
+	err = response.Errors[topic][partitionID]
+	switch err {
+	case ErrNoError:
+		return nil
+	case ErrNotCoordinatorForConsumer:
+		goto RetryCommitHandler
+	default:
+		Logger.Println("Unexpected error for CommitOffset:", err)
+		return err
+	}
+
+RetryCommitHandler:
+	if attemptsRemaining > 0 {
+		client.coordinatorLock.Lock()
+		delete(client.coordinators, consumerGroup)
+		client.coordinatorLock.Unlock()
+
+		time.Sleep(100 * time.Millisecond)
+		return client.tryCommitOffset(consumerGroup, topic, partitionID, offset, metadata, attemptsRemaining-1)
+	} else {
+		return err
+	}
+}
+
+func (client *client) FetchOffset(consumerGroup string, topic string, partitionID int32) (int64, string, error) {
+	return client.tryFetchOffset(consumerGroup, topic, partitionID, 5)
+}
+
+func (client *client) tryFetchOffset(consumerGroup string, topic string, partitionID int32, attemptsRemaining int) (int64, string, error) {
+	var err error
+
+	broker, err := client.getCoordinator(consumerGroup, 10)
+	if err != nil {
+		return -1, "", err
+	}
+
+	offsetFetchRequest := new(OffsetFetchRequest)
+	offsetFetchRequest.ConsumerGroup = consumerGroup
+	offsetFetchRequest.AddPartition(topic, partitionID)
+
+	response, err := broker.FetchOffset(offsetFetchRequest)
+
+	if err != nil {
+		goto RetryFetchHandler
+	} else {
+		block := response.Blocks[topic][partitionID]
+		err = block.Err
+		switch err {
+		case ErrNoError:
+			return block.Offset, block.Metadata, nil
+
+		case ErrNotCoordinatorForConsumer:
+			goto RetryFetchHandler
+
+		default:
+			Logger.Println("Unexpected error for FetchOffset:", err)
+			return -1, "", err
+		}
+	}
+
+RetryFetchHandler:
+	if attemptsRemaining > 0 {
+		client.coordinatorLock.Lock()
+		delete(client.coordinators, consumerGroup)
+		client.coordinatorLock.Unlock()
+
+		time.Sleep(100 * time.Millisecond)
+		return client.tryFetchOffset(consumerGroup, topic, partitionID, attemptsRemaining-1)
+	} else {
+		return -1, "", err
+	}
+}
+
+func (client *client) getCoordinator(consumerGroup string, attemptsRemaining int) (*Broker, error) {
+	client.coordinatorLock.RLock()
+	broker, ok := client.coordinators[consumerGroup]
+	client.coordinatorLock.RUnlock()
+
+	if ok {
+		_ = broker.Open(client.conf)
+		return broker, nil
+	}
+
+	var err error
+
+	request := new(ConsumerMetadataRequest)
```

Comment:
if we're mapping to the existing metadata pattern:
- the rest of this function (from here down) should be it's own `refreshCoordinatorMetadata()` or something, so that `getCoordinator` looks almost exactly like `Leader`
- `getCoordinator` should be public and just called `Coordinator`

Thoughts?